### PR TITLE
Fix various issues with document saving

### LINKED ISSUE #209: Asset Transform Deadlock
image

[Editor.zip](https://github.com/ezEngine/ezEngine/files/4556432/Editor.zip)

Built from 2b745c401cb54778eb14aeaa8e5da11b32de3e90

Steps taken: Create a new asset (collision mesh), transform it manually (CTRL+E). Seems like it deadlocks because it also saves the asset first.

### LINKED ISSUE #246: Crashing when "transforming" loading Sponza resource. Clicked on gear on bottom left of asset that is imported.
Crashes in DocumentTasks.cpp line 23 in function Execute..

It looks like m_document is being freed before calling Execute.

Source of problem is that in MeshImportUtils.cpp line 328. WaitForPendingTasks is causing an Async save. But the Document has already been saved and closed.

Callstack.
 	vcruntime140d.dll!00007ffe2d141392()	Unknown
 	ezFoundation.dll!ezDynamicArrayBase<char>::operator=(ezDynamicArrayBase<char> && rhs) Line 96	C++
 	ezFoundation.dll!ezDynamicArray<char,ezDefaultAllocatorWrapper>::operator=(ezDynamicArray<char,ezDefaultAllocatorWrapper> && rhs) Line 300	C++
 	ezFoundation.dll!ezHybridArray<char,32,ezDefaultAllocatorWrapper>::operator=(ezHybridArray<char,32,ezDefaultAllocatorWrapper> && rhs) Line 52	C++
 	ezFoundation.dll!ezHybridStringBase<32>::operator=(ezHybridStringBase<32> && rhs) Line 119	C++
 	[External Code]	
>	ezToolsFoundation.dll!ezSaveDocumentTask::Execute() Line 23	C++
 	ezFoundation.dll!ezTask::Run(unsigned int uiInvocation) Line 55	C++
 	ezFoundation.dll!ezTaskSystem::ExecuteTask(ezTaskPriority::Enum FirstPriority, ezTaskPriority::Enum LastPriority, bool bOnlyTasksThatNeverWait, const ezTaskGroupID & WaitingForGroup, ezAtomicInteger<int> * pWorkerState) Line 125	C++
 	ezFoundation.dll!ezTaskWorkerThread::Run() Line 67	C++
 	ezFoundation.dll!RunThread(ezThread * pThread) Line 49	C++
 	ezFoundation.dll!ezThreadClassEntryPoint(void * lpThreadParameter) Line 15	C++
 	[External Code]	

Windows 10
Master Branch.

Commit hash:	e4d783469d6a3798e6bb91670f2163e959f37a2b
Parent:			5bc6718511

### PR DESCRIPTION
**Pull request by Sanakan8472**
- Closing a document or re-saving will now block until any pending save is completed.
- Save on manual transform is now performed outside the curator lock.
- Opening a project in editor tests now makes a copy of it to make sure tests run on a clean state.
- Add unit tests.
Resolves #209, Resolves #246

### CODE REVIEW

File: Data/UnitTests/EditorTest/Meshes/Cube.obj [hunk 1]
```diff
@@ -0,0 +1,54 @@
+mtllib Cube.mtl
```
Comment (jankrassnigg) [inline Data/UnitTests/EditorTest/Meshes/Cube.obj:1]: Couldn't you just generate a cube mesh procedurally through the mesh asset ?

File: Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
+#include <EditorTestPCH.h>
+
+#include "AssetDocumentTest.h"
```
Comment (jankrassnigg) [inline Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp]: Include style

File: Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    EZ_TEST_BOOL(pDoc != nullptr);
+    ProcessEvents();
+  }
+
+  ezObjectAccessorBase* pAcc = pDoc->GetObjectAccessor();
+  const ezDocumentObject* pMeshAsset = pDoc->GetObjectManager()->GetRootObject()->GetChildren()[0];
+  EZ_TEST_BLOCK(ezTestBlock::Enabled, "Transform")
+  {
+    pAcc->StartTransaction("Edit Mesh");
+    EZ_TEST_BOOL(pAcc->SetValue(pMeshAsset, "MeshFile", "Meshes/Cube.obj").Succeeded());
```
Comment (jankrassnigg) [inline Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp:127]: Could just set the mesh type to "box" instead.
Unless you really want a file dependency.

File: Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    EZ_TEST_BOOL(res.Succeeded());
+
+    res = ezAssetCurator::GetSingleton()->TransformAsset(pDoc->GetGuid(), ezTransformFlags::ForceTransform | ezTransformFlags::TriggeredManually);
+    EZ_TEST_BOOL(res.Succeeded());
+    ProcessEvents();
+
+    /*  while (true)
+      {
+        ProcessEvents();
+      }*/
```
Comment (jankrassnigg) [inline Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp]: ?

File: Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <EditorTestPCH.h>
+
+#include "../TestClass/TestClass.h"
```
Comment (jankrassnigg) [inline Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.h]: include style

File: Code/UnitTests/EditorTest/TestClass/TestClass.cpp [hunk 6]
```diff
@@ -201,15 +201,39 @@ ezResult ezEditorTest::OpenProject(const char* path)
...
+    ezLog::Error("Failed to delete old project temp folder '{0}'.", absTempPath);
+    return EZ_FAILURE;
+  }
+  if (ezOSFile::CopyFolder(absPath, absTempPath).Failed())
+  {
+    ezLog::Error("Failed to copy project '{0}' to temp location: '{1}'.", absPath, absTempPath);
+    return EZ_FAILURE;
+  }
+
+
```
Comment (jankrassnigg) [inline Code/UnitTests/EditorTest/TestClass/TestClass.cpp:227]: Note that we also have per project + per document state that is saved in the user folder.

Oh, well, I see ezEditorTestApplication::AfterCoreSystemsStartup already enables 'safe' mode, so this is probably enough.

File: Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+class ezEditorAssetDocumentTest : public ezEditorTest
+{
+public:
+  typedef ezEditorTest SUPER;
```
Comment (jankrassnigg) [inline Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.h]: using SUPER =

**Review by jankrassnigg — APPROVED** [review]
Approved with suggestions.

File: Data/UnitTests/EditorTest/Meshes/Cube.obj (hunk 1 above)
Comment (Sanakan8472) [reply Data/UnitTests/EditorTest/Meshes/Cube.obj:1]: Yes absolutely but I want to use the same to test the curator as it is an easy file to modify in code to test file system watcher etc. So just a bit of foundation building on my side.

File: Code/UnitTests/EditorTest/TestClass/TestClass.cpp (hunk 6 above)
Comment (Sanakan8472) [reply Code/UnitTests/EditorTest/TestClass/TestClass.cpp:227]: ezEditorTest::CreateAndLoadProject already dumps the temp projects in the same folder. I could create a subfolder /temp/ ?

File: Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp (hunk 3 above)
Comment (Sanakan8472) [reply Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp:127]: This is needed to run into the code path in the bug: MeshImportUtils.cpp.

Comment (Sanakan8472) [reply Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp:127]: Added a comment to make this clearer.

File: Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp (hunk 4 above)
Comment (Sanakan8472) [reply Code/UnitTests/EditorTest/AssetDocument/AssetDocumentTest.cpp]: wups.

File: Code/UnitTests/EditorTest/TestClass/TestClass.cpp (hunk 6 above)
Comment (jankrassnigg) [reply Code/UnitTests/EditorTest/TestClass/TestClass.cpp:227]: I'm not against it, though I also don't know whether it is necessary. Do as you see fit.

**Pull request merged**